internal/handlers/pg/msg_findandmodify.go

```diff
@@ -110,23 +87,144 @@ func (h *Handler) MsgFindAndModify(ctx context.Context, msg *wire.OpMsg) (*wire.
+		if err != nil {
+			return nil, err
+		}
+
+		var reply wire.OpMsg
+		must.NoError(reply.SetSections(wire.OpMsgSection{
+			Documents: []*types.Document{must.NotFail(types.NewDocument(
+				"lastErrorObject", must.NotFail(types.NewDocument("n", int32(1))),
+				"value", resDocs[0],
+				"ok", float64(1),
+			))},
+		}))
+		return &reply, nil
+	}
+
+	if common.HasUpdateOperator(params.update) {
+		upsert := resDocs[0].DeepCopy()
+		err = common.UpdateDocument(upsert, params.update)
+		if err != nil {
+			return nil, err
+		}
+
+		_, err = h.update(ctx, params.db, params.collection, upsert)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		_, err = h.delete(ctx, resDocs, params.db, params.collection)
+		if err != nil {
+			return nil, err
+		}
+
+		err = h.insert(ctx, params.update, params.db, params.collection)
+		if err != nil {
+			return nil, err
 		}
 	}
 
 	var reply wire.OpMsg
-	err = reply.SetSections(wire.OpMsgSection{
-		Documents: []*types.Document{types.MustNewDocument(
-			"ok", float64(1),
-		)},
-	})
-	if err != nil {
-		return nil, lazyerrors.Error(err)
+	resultDoc := resDocs[0]
+	if params.returnNewDocument {
+		resultDoc = params.update
 	}
+	must.NoError(reply.SetSections(wire.OpMsgSection{
+		Documents: []*types.Document{must.NotFail(types.NewDocument(
+			"lastErrorObject", must.NotFail(types.NewDocument("n", int32(1), "updatedExisting", true)),
+			"value", resultDoc,
+			"ok", float64(1),
+		))},
+	}))
 
 	return &reply, nil
 }
+
+type findAndModifyParams struct {
+	db, collection                    string
+	query, sort, update               *types.Document
+	remove, upsert, returnNewDocument bool
+}
+
+func prepareFindAndModifyParams(document *types.Document, command string) (*findAndModifyParams, error) {
+	var err error
+	var db, collection string
+	if db, err = common.GetRequiredParam[string](document, "$db"); err != nil {
+		return nil, err
+	}
+	if collection, err = common.GetRequiredParam[string](document, command); err != nil {
+		return nil, err
+	}
+
+	if collection == "" {
+		return nil, common.NewErrorMsg(
+			common.ErrInvalidNamespace,
+			fmt.Sprintf("Invalid namespace specified '%s.'", db),
+		)
+	}
+
+	var remove bool
+	if remove, err = common.GetBoolOptionalParam(document, "remove"); err != nil {
+		return nil, err
+	}
+	var returnNewDocument bool
+	if returnNewDocument, err = common.GetBoolOptionalParam(document, "new"); err != nil {
+		return nil, err
+	}
+	var upsert bool
+	if upsert, err = common.GetBoolOptionalParam(document, "upsert"); err != nil {
+		return nil, err
+	}
+
+	var query *types.Document
+	if query, err = common.GetOptionalParam(document, "query", query); err != nil {
+		return nil, err
+	}
+
+	var sort *types.Document
+	if sort, err = common.GetOptionalParam(document, "sort", sort); err != nil {
+		return nil, err
+	}
+
+	var update *types.Document
+	updateParam, err := document.Get("update")
+	if err != nil && !remove {
```

Comment:
if err != nil 
or
update && remove

isn't it? Could you please add test for it?

Reply:
There is a test for that → "NotEnoughParameters".